src/java/com/rapleaf/jack/queries/GenericQuery.java

```diff
@@ -0,0 +1,288 @@
+      record.addColumn(column, value);
+    }
+    return record;
+  }
+
+  private String getQueryStatement() {
+    return getSelectClause()
+        + getFromClause()
+        + getJoinClause()
+        + getWhereClause()
+        + getGroupByClause()
+        + getOrderClause()
+        + getLimitClause();
+  }
+
+  private String getSelectClause() {
+    if (!groupByColumns.isEmpty()) {
+      if (selectedColumns.isEmpty()) {
+        throw new RuntimeException("The SELECT list cannot be empty when the GROUP BY clause is specified.");
+      }
+
+      for (Column column : selectedColumns) {
+        if (!groupByColumns.contains(column) && !(column instanceof AggregatedColumn)) {
+          throw new RuntimeException("The non-aggregated column " + column.getSqlKeyword() +
+              " not named in the GROUP BY clause cannot be included in the SELECT list.");
+        }
+      }
+    }
+
+    if (selectedColumns.isEmpty()) {
+      for (Table table : includedTables) {
+        selectedColumns.addAll(table.getAllColumns());
+      }
+    }
+    return getClauseFromColumns(selectedColumns, "SELECT ", ", ");
+  }
+
+  private String getFromClause() {
+    return "FROM " + includedTables.get(0).getSqlKeyword() + " ";
+  }
+
+  private String getJoinClause() {
+    return getClauseFromQueryConditions(joinConditions, "", " ");
+  }
+
+  private String getWhereClause() {
+    return getClauseFromQueryConditions(whereConstraints, "WHERE ", " AND ");
+  }
+
+  private String getGroupByClause() {
+    return getClauseFromColumns(groupByColumns, "GROUP BY ", ", ");
+  }
+
+  private String getOrderClause() {
+    if (orderCriteria.isEmpty()) {
+      return "";
+    } else {
+      return getClauseFromQueryConditions(orderCriteria, "ORDER BY ", ", ");
+    }
+  }
+
+  private String getLimitClause() {
+    if (limitCriteria.isPresent()) {
+      return limitCriteria.get().getSqlStatement() + " ";
+    } else {
+      return "";
+    }
+  }
+
+  private String getClauseFromColumns(Collection<Column> columns, String initialKeyword, String separator) {
+    if (columns.isEmpty()) {
+      return "";
+    }
+
+    StringBuilder clause = new StringBuilder(initialKeyword);
+    Iterator<Column> it = columns.iterator();
+    while (it.hasNext()) {
+      clause.append(it.next().getSqlKeyword());
+      if (it.hasNext()) {
+        clause.append(separator);
+      }
+    }
+
+    return clause.append(" ").toString();
+  }
+
+  private <T extends QueryCondition> String getClauseFromQueryConditions(Collection<T> conditions, String initialKeyword, String separator) {
+    if (conditions.isEmpty()) {
+      return "";
+    }
+
+    StringBuilder clause = new StringBuilder(initialKeyword);
+    Iterator<T> it = conditions.iterator();
+    while (it.hasNext()) {
+      clause.append(it.next().getSqlStatement());
+      if (it.hasNext()) {
+        clause.append(separator);
+      }
+    }
+
+    return clause.append(" ").toString();
+  }
+}
```

Comment:
`getClauseFromColumns` and `getClauseFromQueryConditions` are basically the same except for the type of elements in the collection. I thought about making `Column` and other query condition classes implement the same interface with a `getSqlStatement` method to unify them, but that involves other namings issues. I'd like to leave that to a sweep ramp ticket.